src/move/research/prover-examples/sources/main.move

```diff
@@ -16,10 +16,31 @@ module prover_examples::main {
         invariant value <= MAX_VALUE;
     }
 
+    // Ensure that wherever a PhantomKeyStruct<T> exists, the account_address field is equal to the
+    // account address of the struct.
     invariant<T> forall account_address: address where exists<PhantomKeyStruct<T>>(
         account_address
     ): global<PhantomKeyStruct<T>>(account_address).account_address == account_address;
 
+    // Ensure that on update to a PhantomKeyStruct<T>, the value field is incremented by 1, or reset
+    // to 0 if it was already at MAX_VALUE. A global representation of the spec for
+    // increment_phantom_key_struct_with_rollover.
+    invariant<T> update[global] forall account_address: address where exists<
+        PhantomKeyStruct<T>>(account_address)
+        && old(
+            exists<PhantomKeyStruct<T>>(account_address)
+        ):
+        global<PhantomKeyStruct<T>>(account_address).value
+            == old(
+                global<PhantomKeyStruct<T>>(account_address).value
+            ) + 1
+            || (
+                global<PhantomKeyStruct<T>>(account_address).value == 0
+                    && old(
+                        global<PhantomKeyStruct<T>>(account_address).value
+                    ) == MAX_VALUE
+            );
```

Comment:
@wrwg any reason why I'd get the error `global memory invariant does not hold`?

This is essentially identical to the `ensures` statement from the corresponding function spec, which passes

Reply:
hi @alnoki, you need to specify in the precondition that the value must be changed:
```
    invariant update forall account_address: address where old(
            exists<PhantomKeyStruct>(account_address)
        ) &&  exists<PhantomKeyStruct>(account_address) && 
        global<PhantomKeyStruct>(account_address).value != old(global<PhantomKeyStruct>(account_address).value):
        (old(global<PhantomKeyStruct>(account_address).value) == MAX_VALUE && 
        global<PhantomKeyStruct>(account_address).value == 0) || 
        global<PhantomKeyStruct>(account_address).value == old(global<PhantomKeyStruct>(account_address).value) + 1;
```
Otherwise, the property will not hold for addresses that are not touched

Reply:
@rahxephon89 thanks for clarifying, I've addressed this in 8328204 however I am still a bit confused

If it already has `update[global]`, then why do I need to specify an update again in the precondition?

Reply:
That is because the property is quantified over all addresses that have the resource `PhantomKeyStruct` and it is being checked whenever there is an update on any one of the addresses. If we don't add this precondition, resource under other addresses don't satisfy this property